File: backend/app/services/ai_service.py
from bson import ObjectId
from google import genai
from app.config import GEMINI_API_KEY
from app.services.prompt_builder import build_prompt
from app.utils.json_parser import parse_ai_json
from app.database import users_collection, wardrobe_collection

client = genai.Client(api_key=GEMINI_API_KEY)


def generate_outfit_response(user_id: str, message: str):

    # 🔹 Validate ObjectId
    try:
        object_id = ObjectId(user_id)
    except Exception:
        return {"error": "Invalid user ID format"}

    # 🔹 Fetch user
    user = users_collection.find_one({"_id": object_id})
    if not user:
        return {"error": "User not found"}

    # 🔹 Fetch wardrobe
    wardrobe_items = wardrobe_collection.find({"user_id": object_id})

    wardrobe = [
        {
            "category": item.get("category"),
            "name": item.get("name"),
            "color": item.get("color"),
            "style": item.get("style")
        }
        for item in wardrobe_items
    ]

    if not wardrobe:
        wardrobe = [{"category": "Shirt", "name": "Basic black shirt", "color": "Black", "style": "Casual"}]

    # 🔹 Safe profile mapping
    profile = {
        "gender": user.get("gender", "male"),
        "body_type": user.get("body_type", "fit"),
        "style": ", ".join(user.get("style_preferences", ["casual"])),
    }

    prompt = build_prompt(profile, wardrobe, message)

    try:
        response = client.models.generate_content(
            model="gemini-2.5-flash",
            contents=prompt,
            config={
                "temperature": 0.3,
                "max_output_tokens": 1200
            }
        )

        parsed_json = parse_ai_json(response.text)
        return parsed_json

    except Exception as e:
        return {
            "error": "AI generation failed",
            "details": str(e)
        }

File: backend-node/src/services/ai.service.js
import { GoogleGenerativeAI } from "@google/generative-ai";

import { GEMINI_API_KEY } from "../config/config.js";
import User from "../models/User.js";
import Wardrobe from "../models/Wardrobe.js";

import buildPrompt from "./prompt.builder.js";
import parseAIJson from "../utils/jsonParser.js";

const genAI = new GoogleGenerativeAI(GEMINI_API_KEY);

const model = genAI.getGenerativeModel({
  model: "gemini-2.5-flash"
});

const generateOutfitResponse = async (user_id, message) => {

  try {

    const user = await User.findById(user_id);

    if (!user) {
      return { error: "User not found" };
    }

    const wardrobeItems = await Wardrobe.find({ user_id });

    let wardrobe = wardrobeItems.map(item => ({
      category: item.category,
      name: item.name,
      color: item.color,
      style: item.style
    }));

    if (!wardrobe.length) {

      wardrobe = [{
        category: "Shirt",
        name: "Basic black shirt",
        color: "Black",
        style: "Casual"
      }];
    }

    const profile = {
      gender: user.gender || "male",
      body_type: user.body_type || "fit",
      style: (user.style_preferences || ["casual"]).join(", ")
    };

    const prompt = buildPrompt(profile, wardrobe, message);

    const result = await model.generateContent(prompt);

    const text = result.response.text();

    return parseAIJson(text);

  } catch (error) {

    return {
      error: "AI generation failed",
      details: error.message
    };
  }
};

export default generateOutfitResponse;